src/memray/_memray/hooks.cpp

```diff
@@ -304,19 +306,76 @@ posix_memalign(void** memptr, size_t alignment, size_t size) noexcept
+// loaded in the process memory space. This is needed so we can correctly track
+// allocations in those libraries by overriding their PLT entries and also so we
+// can properly map the addresses of the symbols in those libraries when we
+// resolve later native traces. Unfortunately, we can't just override dlopen
+// directly because of the following edge case: when a shared library dlopen's
+// another by name (e.g. dlopen("libfoo.so")), the dlopen call will honor the
+// RPATH/RUNPATH of the calling library if it's set. Some libraries set an
+// RPATH/RUNPATH based on $ORIGIN (the path of the calling library) to load
+// dependencies from a relative directory based on the location of the calling
+// library. This means that if we override dlopen, we'll end up loading the
+// library from the wrong path or more likely, not loading it at all because the
+// dynamic loader will think the memray extenion it's the calling library and
+// the RPATH of the real calling library will not be honoured.
+//
+// To work around this, we override dlsym instead and override the symbols in
+// the loaded libraries only the first time we have seen a handle passed to
+// dlsym. This works because for a symbol from a given dlopen-ed library to
+// appear in a call stack, *something* from that library has to be dlsym-ed
+// first. The only exception to this are static initializers, but we cannot
+// track those anyway by overriding dlopen as they run within the dlopen call
+// itself.
+
+class DlsymCache
+{
+  public:
+    void addEntry(const void* handle)
+    {
+        std::unique_lock lock(mutex_);
+        d_handles.insert(handle);
+    }
+
+    bool lookup(const void* handle) const
+    {
+        std::shared_lock lock(mutex_);
+        auto it = d_handles.find(handle);
+        if (it != d_handles.end()) {
+            return true;
+        }
+        return false;
+    }
+
+    void deleteEntry(const void* handle)
+    {
+        std::unique_lock lock(mutex_);
+        d_handles.erase(handle);
+    }
+
+  private:
+    mutable std::shared_mutex mutex_;
+    std::set<const void*> d_handles;
+};
+
+static DlsymCache dlsym_cache;
+
 void*
-dlopen(const char* filename, int flag) noexcept
+dlsym(void* handle, const char* symbol) noexcept
 {
-    assert(hooks::dlopen);
+    assert(hooks::dlsym);
     void* ret;
     {
         tracking_api::RecursionGuard guard;
-        ret = hooks::dlopen(filename, flag);
+        ret = hooks::dlsym(handle, symbol);
     }
-    if (ret) {
+    if (ret && !dlsym_cache.lookup(handle)) {
+        dlsym_cache.addEntry(handle);
         tracking_api::Tracker::invalidate_module_cache();
-        if (filename
-            && (nullptr != strstr(filename, "/_greenlet.") || nullptr != strstr(filename, "/greenlet.")))
+        if (symbol
+            && (nullptr != strstr(symbol, "PyInit_greenlet")
+                || nullptr != strstr(symbol, "PyInit__greenlet")))
```

Comment:
This is gross, but it might actually be slightly less gross than the original version, heh.

But don't we want strcmp instead of strstr?
```suggestion
            && (0 == strcmp(symbol, "PyInit_greenlet")
                || 0 == strcmp(symbol, "PyInit__greenlet")))
```